File: challenge_workbook/dp_acm231.py
#bag question
#recursion solution and DP solution
n = 4
w,v = [2,1,3,2],[3,2,4,2]
W = 5
res = 0
#i: no. of decision made, j: remained weight
dp = [[-1 for i in range(W+1)] for j in range(n+1)]
'''for i in range(n+1):
    dp.append([])
    for j in range(W+1):
        dp[i].append(-1)
'''
#recursion method:
def rec(i,j):
    if dp[i][j] >= 0:
        return dp[i][j]
    
    
    if (i == n):
        res = 0
        
        
    elif (j < w[i]):
        res = rec(i+1, j)
        
    else:
        res  = max(rec(i+1, j), rec(i+1, j-w[i])+v[i])
    dp[i][j] = res
    return  res

# DP method
dp = [[0 for i in range(W+1)] for j in range(n+1)]
def dpSolve():
    for i in range(n-1, -1, -1):
        for j in range(0, W+1):
            if (j < w[i]):
                dp[i][j] = dp[i+1][j]
            else:
                dp[i][j] = max(dp[i+1][j], dp[i+1][j-w[i]]+v[i])
#Test
r = rec(0,W)
print(dp)
dpr = dpSolve()
print(dp)

File: challenge_workbook/dp_acm231.cpp
#include<iostream>
#include<cstring>
using namespace std;

int n = 4, W = 5;
int w[100] = {2, 1, 3, 2}, v[100] = {3,2,4,2};


int dp[5][6]; 
int rec(int i, int j){
    if (dp[i][j] >=0){
        return dp[i][j];
    }
    int res;
    if (i == n){
        res = 0 ;
    }
    else if(j < w[i]){
        res = rec(i+1, j);
    }else{
        res = max(rec(i+1,j), rec(i+1, j-w[i])+v[i]);
    }
    dp[i][j] = res;
    return res;
}
void solve(){
    memset(dp, -1, sizeof(dp));
    rec(0,W);
    //printf("%d",rec(0,W));
    for (int i = 0; i < 5;i ++){
        for (int j = 0; j < 6; j ++)
        cout << dp[i][j] << " ";
    }
}
int main(){
    solve();
    return 0;
}